bpf/flows_sock.c

```diff
@@ -0,0 +1,254 @@
+        __builtin_memcpy(id->dst_ip.s6_addr + sizeof(ip4in6), &daddr, sizeof(daddr));
+
+        hdr_len = ETH_HLEN + hdr_len;
+        break;
+    }
+    case ETH_P_IPV6:
+        bpf_skb_load_bytes(skb, ETH_HLEN + offsetof(struct ipv6hdr, nexthdr), &proto, sizeof(proto));
+
+        bpf_skb_load_bytes(skb, ETH_HLEN + offsetof(struct ipv6hdr, saddr), &id->src_ip.s6_addr, sizeof(id->src_ip.s6_addr));
+        bpf_skb_load_bytes(skb, ETH_HLEN + offsetof(struct ipv6hdr, daddr), &id->dst_ip.s6_addr, sizeof(id->dst_ip.s6_addr));
+
+        hdr_len = ETH_HLEN + sizeof(struct ipv6hdr);
+        break;
+    default:
+        return false;
+    }
+
+    id->src_port = 0;
+    id->dst_port = 0;
+    id->transport_protocol = proto;
+
+    switch(proto) {
+        case IPPROTO_TCP: {
+            u16 port;
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __tcphdr, source), &port, sizeof(port));
+            id->src_port = __bpf_htons(port);
+
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __tcphdr, dest), &port, sizeof(port));
+            id->dst_port = __bpf_htons(port);
+
+            u8 doff;
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __tcphdr, ack_seq) + 4, &doff, sizeof(doff)); // read the first byte past __tcphdr->ack_seq, we can't do offsetof bit fields
+            doff &= 0xf0; // clean-up res1
+            doff >>= 4; // move the upper 4 bits to low
+            doff *= 4; // convert to bytes length
+
+            hdr_len += doff;
+
+            u8 flags;
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __tcphdr, ack_seq) + 4 + 1, &flags, sizeof(flags)); // read the second byte past __tcphdr->doff, again bit fields offsets
+            *custom_flags = ((u16)flags & 0x00ff);
+
+            if ((skb->len - hdr_len) < 0) { // less than 0 is a packet we can't parse
+                return false;
+            }
+
+            break;
+        } 
+        case IPPROTO_UDP: {
+            u16 port;
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __udphdr, source), &port, sizeof(port));
+            id->src_port = __bpf_htons(port);
+            bpf_skb_load_bytes(skb, hdr_len + offsetof(struct __udphdr, dest), &port, sizeof(port));
+            id->dst_port = __bpf_htons(port);
+        }
+    }
+
+    // custom flags
+    if ((*custom_flags & (TCPHDR_ACK | TCPHDR_SYN))) {
+        *custom_flags |= SYN_ACK_FLAG;
+    } else if ((*custom_flags & (TCPHDR_ACK | TCPHDR_FIN))) {
+        *custom_flags |= FIN_ACK_FLAG;
+    } else if ((*custom_flags & (TCPHDR_ACK | TCPHDR_RST))) {
+        *custom_flags |= RST_ACK_FLAG;
+    }
+
+    return true;
+}
+
+static __always_inline bool same_ip(u8 *ip1, u8 *ip2) {
+    for (int i=0; i<16; i+=4) {
+        if (*((u32 *)(ip1+i)) != *((u32 *)(ip2+i))) {
+            return false;
+        }
+    }
+    
+    return true;
+}
+
+SEC("socket/http_filter")
+int socket__http_filter(struct __sk_buff *skb) {
```

Comment:
Even though this is __sk_buff, I couldn't directly reuse the code for TC, skb->data and skb->data_end don't work in socket filters. I had to use skb_load_bytes everywhere to read the data. Maybe there's a trick to make it work, but I don't know of one.